File: client/src/types/coupon.ts
// ── Shared label maps ────────────────────────────────────────────────────────

export const CATEGORY_LABELS: Record<number, string> = {
  0: 'לא ידוע', 1: 'מזון', 2: 'אופנה', 3: 'נסיעות',
  4: 'בריאות', 5: 'בידור', 6: 'סופרמרקט', 7: 'רב-מותגי',
  8: 'תחנת דלק', 9: 'כרטיס מתנה', 99: 'אחר',
}

export const EXPIRY_TYPE_LABELS: Record<number, string> = {
  0: 'לא ידוע', 1: 'תאריך מדויק', 2: 'עד הודעה חדשה', 3: 'משך יחסי',
}

export const STATUS_LABELS: Record<number, string> = {
  0: 'לא ידוע', 1: 'פעיל', 2: 'פג תוקף', 3: 'מומש',
}

// ── Categories ───────────────────────────────────────────────────────────────

export interface CategorySummary {
  category: number
  displayName: string
  totalCoupons: number
  activeCoupons: number
  totalActiveRemainingAmount: number | null
}

// ── Analyze ──────────────────────────────────────────────────────────────────

// Request sent to POST /api/coupons/analyze
export interface AnalyzeCouponRequest {
  rawText: string
}

// Response from POST /api/coupons/analyze
// Enum fields (category, expirationType, status) are numeric values from the backend.
export interface CouponAnalysisResult {
  title: string | null
  provider: string | null
  merchantName: string | null
  normalizedMerchantName: string | null
  category: number
  originalAmount: number | null
  remainingAmount: number | null
  currency: string | null
  couponCode: string | null
  numerator: string | null
  voucherUrl: string | null
  expirationDate: string | null   // ISO 8601 date string or null
  expirationText: string | null
  expirationType: number
  onlineRedeemable: boolean
  status: number
  confidence: number              // 0.0–1.0
  conditionsText: string | null
  warnings: string[]
}

// ── Save / CRUD ───────────────────────────────────────────────────────────────

// Request sent to POST /api/coupons
export interface CreateCouponRequest {
  rawText: string
  title: string | null
  provider: string | null
  merchantName: string | null
  normalizedMerchantName: string | null
  category: number
  originalAmount: number | null
  remainingAmount: number | null
  currency: string | null
  couponCode: string | null
  numerator: string | null
  voucherUrl: string | null
  expirationDate: string | null
  expirationText: string | null
  expirationType: number
  onlineRedeemable: boolean
  status: number
  confidence: number
  conditionsText: string | null
}

// Response from GET /api/coupons and GET /api/coupons/{id}
export interface CouponDto {
  id: number
  rawText: string
  title: string | null
  provider: string | null
  merchantName: string | null
  normalizedMerchantName: string | null
  category: number
  originalAmount: number | null
  remainingAmount: number | null
  currency: string | null
  couponCode: string | null
  numerator: string | null
  voucherUrl: string | null
  expirationDate: string | null
  expirationText: string | null
  expirationType: number
  onlineRedeemable: boolean
  status: number
  confidence: number
  conditionsText: string | null
  receivedAt: string | null
  createdAt: string
  updatedAt: string
}

File: server/Models/Coupon.cs
using CouponManager.Api.Enums;

namespace CouponManager.Api.Models;

public class Coupon
{
    public int Id { get; set; }

    // The original raw SMS/message text pasted by the user
    public string RawText { get; set; } = string.Empty;

    // Human-readable title, either extracted or user-edited
    public string? Title { get; set; }

    // Who issued the coupon (e.g. "Pluxee", "Multipass")
    public string? Provider { get; set; }

    // Merchant name as found in the raw text
    public string? MerchantName { get; set; }

    // Canonical merchant name after normalization
    public string? NormalizedMerchantName { get; set; }

    public CouponCategory Category { get; set; } = CouponCategory.Unknown;

    // Face value at the time of issue
    public decimal? OriginalAmount { get; set; }

    // Remaining usable value (may differ after partial redemption)
    public decimal? RemainingAmount { get; set; }

    // Currency code, e.g. "ILS", "USD"
    public string? Currency { get; set; }

    // Alphanumeric redemption code
    public string? CouponCode { get; set; }

    // "1 of 3 uses" style text
    public string? Numerator { get; set; }

    // Redemption URL (fake/placeholder in dev)
    public string? VoucherUrl { get; set; }

    // Parsed expiration date
    public DateTime? ExpirationDate { get; set; }

    // Raw expiration string from the message (e.g. "עד 31.12.25")
    public string? ExpirationText { get; set; }

    public ExpirationType ExpirationType { get; set; } = ExpirationType.Unknown;

    // Whether the coupon can be used online
    public bool OnlineRedeemable { get; set; }

    public CouponStatus Status { get; set; } = CouponStatus.Active;

    // Parser confidence score (0.0–1.0)
    public double Confidence { get; set; }

    // Raw conditions paragraph from the message
    public string? ConditionsText { get; set; }

    // When the original message was reportedly sent/received
    public DateTime? ReceivedAt { get; set; }

    public DateTime CreatedAt { get; set; } = DateTime.UtcNow;

    public DateTime UpdatedAt { get; set; } = DateTime.UtcNow;
}
